toolchain/sem_ir/builtin_function_kind.cpp

```diff
@@ -0,0 +1,179 @@
+// specifies that the builtin takes values of two integer types and returns a
+// value of a third integer type. Types used within the signature should provide
+// a `Check` function that validates that the Carbon type is expected:
+//
+//   auto Check(const File&, ValidateState&, TypeId) -> bool;
+//
+// To constrain that the same type is used in multiple places in the signature,
+// `TypeParam<I, T>` can be used. For example:
+//
+// `auto (TypeParam<0, AnyInt>, AnyInt) -> TypeParam<0, AnyInt>`
+//
+// describes a builtin that takes two integers, and whose return type matches
+// its first parameter type. For convenience, typedefs for `TypeParam<I, T>`
+// are used in the descriptions of the builtins.
+template <typename SignatureFnType>
+static auto ValidateSignature(const File& sem_ir,
+                              llvm::ArrayRef<TypeId> arg_types,
+                              TypeId return_type) -> bool {
+  using SignatureTraits = llvm::function_traits<SignatureFnType*>;
+  ValidateState state;
+
+  // Must have expected number of arguments.
+  if (arg_types.size() != SignatureTraits::num_args) {
+    return false;
+  }
+
+  // Argument types must match.
+  if (![&]<std::size_t... Indexes>(std::index_sequence<Indexes...>) {
+        return ((SignatureTraits::template arg_t<Indexes>::Check(
+                    sem_ir, state, arg_types[Indexes])) &&
+                ...);
+      }(std::make_index_sequence<SignatureTraits::num_args>())) {
+    return false;
+  }
+
+  // Result type must match.
+  if (!SignatureTraits::result_t::Check(sem_ir, state, return_type)) {
+    return false;
+  }
+
+  return true;
+}
+
+// Descriptions of builtin functions follow. For each builtin, a corresponding
+// `BuiltinInfo` constant is declared describing properties of that builtin.
+namespace BuiltinFunctionInfo {
+
+// Convenience name used in the builtin type signatures below for a first
+// generic type parameter that is constrained to be an integer type.
+using IntT = TypeParam<0, AnyInt>;
+
+// Not a builtin function.
+constexpr BuiltinInfo None = {"", nullptr};
+
+// "int.add": integer addition.
+constexpr BuiltinInfo IntAdd = {"int.add",
+                                ValidateSignature<auto(IntT, IntT)->IntT>};
+
+}  // namespace BuiltinFunctionInfo
+
+CARBON_DEFINE_ENUM_CLASS_NAMES(BuiltinFunctionKind) = {
+#define CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(Name) \
+  BuiltinFunctionInfo::Name.name,
+#include "toolchain/sem_ir/builtin_function_kind.def"
+};
+
+// Returns the builtin function kind with the given name, or None if the name
+// is unknown.
+auto BuiltinFunctionKind::ForBuiltinName(llvm::StringRef name)
+    -> BuiltinFunctionKind {
+  return llvm::StringSwitch<BuiltinFunctionKind>(name)
+#define CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(Name) \
+  .Case(BuiltinFunctionInfo::Name.name, BuiltinFunctionKind::Name)
+#include "toolchain/sem_ir/builtin_function_kind.def"
+      .Default(BuiltinFunctionKind::None);
```

Comment:
```suggestion
#define CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(Name) \
  if (name == BuiltinFunctionInfo::Name.name) { return BuiltinFunctionKind::Name; }
#include "toolchain/sem_ir/builtin_function_kind.def"
  return BuiltinFunctionKind::None;
```

How much does StringSwitch buy here? I think this may be similar, but with less to optimize. But maybe I've misread what's going on there.

Reply:
I think this is basically what `StringSwitch` does under the hood, so it's probably not buying us anything.